### PR TITLE
Handle reproducing symlinks for existing links

### PR DESCRIPTION
**Pull request by agirault**
Solves #12

### CODE REVIEW

**Comment by arcivanov** [comment]
That's wonderful, thank you! Can you please add a reproducible test case?

File: src/main/python/wheel_axle/bdist_axle/_file_utils.py [hunk 1]
```diff
@@ -38,7 +38,11 @@ def copy_link(src, dst, update=0, verbose=1, dry_run=0, reproduce_link=False):
                  dir if os.path.basename(dst) == os.path.basename(src) else dst)
 
     if reproduce_link:
-        os.symlink(link_dest, dst, link_dest_isdir)
+        try:
+            os.symlink(link_dest, dst, link_dest_isdir)
+        except FileExistsError:
+            os.remove(dst)
```
Comment (arcivanov) [inline src/main/python/wheel_axle/bdist_axle/_file_utils.py:44]: `os.unlink` please

**Comment by arcivanov** [comment]
> since [super().copy_file](https://github.com/karellen/wheel-axle/blob/master/src/main/python/wheel_axle/bdist_axle/__init__.py#L172-L174) handles it gracefully (no error, just returns false in the return value for "not copied" if it already exists).

> However, symlinks (like prefix/foo.so.0 and prefix/foo.so) are handled by [copy_link](https://github.com/karellen/wheel-axle/blob/master/src/main/python/wheel_axle/bdist_axle/__init__.py#L169) which [seems to assume always doing the copy](https://github.com/karellen/wheel-axle/blob/master/src/main/python/wheel_axle/bdist_axle/__init__.py#L170). Due to the duplicate copies from the separate packages listed above, the second copy fails

I think the safer way is to not remove the target symlink but to continue in a 'false in the return value for not copied"' fashion to replicate the behavior.

File: src/main/python/wheel_axle/bdist_axle/_file_utils.py (hunk 1 above)
Comment (agirault) [reply src/main/python/wheel_axle/bdist_axle/_file_utils.py]: done

Comment (arcivanov) [reply src/main/python/wheel_axle/bdist_axle/_file_utils.py]: Thanks. See my comment below though - I think it's safer to replicate the "copy" behavior - skip if exists. I would also prefer if a warning could be printed that file already exists.

**Comment by agirault** [comment]
> I think the safer way is to not remove the target symlink but to continue in a 'false in the return value for not copied"' fashion to replicate the behavior.

Looking at `super().copy_file`, it's calling `distutils.file_util._copy_file_contents` which does remove the existing content to replace it here: https://github.com/pypa/distutils/blob/main/distutils/file_util.py#L33

The mechanism that avoids copies is using `update` like here: https://github.com/pypa/distutils/blob/main/distutils/file_util.py#L125

It is a parameter of `copy_link`, but it's unused: https://github.com/karellen/wheel-axle/blob/master/src/main/python/wheel_axle/bdist_axle/_file_utils.py#L23

**Comment by arcivanov** [comment]
Very well, then your solution is perfectly fine. Could you be so kind as to add a reproducible test? 

You can use https://github.com/karellen/wheel-axle/blob/master/src/integrationtest/python/build_axle_tests.py as a template and store artifacts similar to https://github.com/karellen/wheel-axle/tree/master/src/integrationtest/resources/test_axle_1.

You could either add duplicate symlinks to an existing test or create a separate test altogether.

Thank you so much!

**Comment by agirault** [comment]
@arcivanov I pushed an additional commit that handles `update` and `dry-run`.

I'll need more time to update a test.

**Comment by arcivanov** [comment]
> @arcivanov I pushed an additional commit that handles `update` and `dry-run`.
> 
> I'll need more time to update a test.

Thanks but handling update and dry-run is not necessary AFAIR. These functions were copied out of the original in distutils/setuptools and behavior was kept exactly the same as the copy file would be except for handling the links. 

I think the safest way would be to go with your original approach limiting it to https://github.com/karellen/wheel-axle/commit/7376aeeb6b7925c39b6db58bd8f4f09df3257819

**Comment by agirault** [comment]
Having support for `update` would have actually prevented the bug, since it would have seen a new file is there already and a new copy isn't necessary.

Re: test -- I found out when looking at the existing test resources that the error does not occur when using `packages=find_packages("src")` instead of `include_package_data=True`, since the nested packages (that cause a duplicate list of files to appear with nesting) don't appear then. It still seems to me that we'd want things to work with `include_package_data`, would you agree?

**Comment by arcivanov** [comment]
> Having support for `update` would have actually prevented the bug, since it would have seen a new file is there already and a new copy isn't necessary.

I'm not sure `update` and `dry_run` are actually used in any of the call paths that I have observed when I debugged it.

**Comment by arcivanov** [comment]
I fixed the issue with a reproducible case in #14. Please test with `pip install -U --pre wheel-axle`.

**Pull request closed without merging**